subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultBuildLogicBuilder.java

```diff
@@ -68,7 +69,9 @@ private List<TaskIdentifier.TaskBasedTaskIdentifier> taskIdentifiersForBuildDepe
         List<TaskIdentifier.TaskBasedTaskIdentifier> tasksToBuild = new ArrayList<>();
         for (Task task : getDependenciesForInternalUse(classpath)) {
             BuildState targetBuild = owningBuildOf(task);
-            assert targetBuild != currentBuild;
+            if (targetBuild == currentBuild) {
```

Comment:
One of the tests I wrote triggered the former assertion. It seems the assumption this code made did not hold true. 

It is probably not right to make this an exception without a deprecation cycle, as it seems possible to trigger this condition in user code currently.

Reply:
* We should test if the existing behavior runs the task or not
* If the task did not run, we can throw this exception now
* If the task did run, (but how the heck would that happen), then we still throw an exception. This really should not happen.

Reply:
This fails with a different exception if we remove the assertion and new exception. 

Leaving this as-is